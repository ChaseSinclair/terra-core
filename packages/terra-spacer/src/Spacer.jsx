import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import { withBps } from 'react-bps';
import 'terra-base/lib/baseStyles';
import styles from './Spacer.scss';

const cx = classNames.bind(styles);

const propTypes = {
  /**
   * Child Nodes.
   */
  children: PropTypes.node,
  /**
   * Sets the margin-top-width. One of 'none', 'small-2', 'small-1', 'small', 'medium', 'large', 'large+1', 'large+2', 'large+3', 'large+4'.
   */
  marginTop: PropTypes.oneOf(['none', 'small-2', 'small-1', 'small', 'medium', 'large', 'large+1', 'large+2', 'large+3', 'large+4']),
  /**
   * Sets the margin-bottom-width. One of 'none', 'small-2', 'small-1', 'small', 'medium', 'large', 'large+1', 'large+2', 'large+3', 'large+4'.
   */
  marginBottom: PropTypes.oneOf(['none', 'small-2', 'small-1', 'small', 'medium', 'large', 'large+1', 'large+2', 'large+3', 'large+4']),
  /**
   * Sets the margin-left-width. One of 'none', 'small-2', 'small-1', 'small', 'medium', 'large', 'large+1', 'large+2', 'large+3', 'large+4'.
   */
  marginLeft: PropTypes.oneOf(['none', 'small-2', 'small-1', 'small', 'medium', 'large', 'large+1', 'large+2', 'large+3', 'large+4']),
  /**
   * Sets the margin-right-width. One of 'none', 'small-2', 'small-1', 'small', 'medium', 'large', 'large+1', 'large+2', 'large+3', 'large+4'.
   */
  marginRight: PropTypes.oneOf(['none', 'small-2', 'small-1', 'small', 'medium', 'large', 'large+1', 'large+2', 'large+3', 'large+4']),
  /**
   * Sets the padding-top-width. One of 'none', 'small-2', 'small-1', 'small', 'medium', 'large', 'large+1', 'large+2', 'large+3', 'large+4'.
   */
  paddingTop: PropTypes.oneOf(['none', 'small-2', 'small-1', 'small', 'medium', 'large', 'large+1', 'large+2', 'large+3', 'large+4']),
  /**
   * Sets the padding-bottom-width. One of 'none', 'small-2', 'small-1', 'small', 'medium', 'large', 'large+1', 'large+2', 'large+3', 'large+4'.
   */
  paddingBottom: PropTypes.oneOf(['none', 'small-2', 'small-1', 'small', 'medium', 'large', 'large+1', 'large+2', 'large+3', 'large+4']),
  /**
   * Sets the padding-left-width. One of 'none', 'small-2', 'small-1', 'small', 'medium', 'large', 'large+1', 'large+2', 'large+3', 'large+4'.
   */
  paddingLeft: PropTypes.oneOf(['none', 'small-2', 'small-1', 'small', 'medium', 'large', 'large+1', 'large+2', 'large+3', 'large+4']),
  /**
   * Sets the padding-right-width. One of 'none', 'small-2', 'small-1', 'small', 'medium', 'large', 'large+1', 'large+2', 'large+3', 'large+4'.
   */
  paddingRight: PropTypes.oneOf(['none', 'small-2', 'small-1', 'small', 'medium', 'large', 'large+1', 'large+2', 'large+3', 'large+4']),
  /**
   * Sets the display to be inline-block
   */
  isInlineBlock: PropTypes.bool,
};

const defaultProps = {
  children: null,
  marginTop: 'none',
  marginBottom: 'none',
  marginRight: 'none',
  marginLeft: 'none',
  paddingTop: 'none',
  paddingBottom: 'none',
  paddingLeft: 'none',
  paddingRight: 'none',
  isInlineBlock: 'false',
};

const Spacer = ({
  marginTop,
  marginBottom,
  marginLeft,
  marginRight,
  paddingTop,
  paddingBottom,
  paddingLeft,
  paddingRight,
  isInlineBlock,
  ...customProps
 }) => {
  const attributes = Object.assign({}, customProps);
  const SpacerClassNames = cx([
    'spacer',
    { 'margin-left-none': marginLeft === 'none' },
    { 'margin-left-small-2': marginLeft === 'small-2' },
    { 'margin-left-small-1': marginLeft === 'small-1' },
    { 'margin-left-small': marginLeft === 'small' },
    { 'margin-left-medium': marginLeft === 'medium' },
    { 'margin-left-large': marginLeft === 'large' },
    { 'margin-left-large+1': marginLeft === 'large+1' },
    { 'margin-left-large+2': marginLeft === 'large+2' },
    { 'margin-left-large+3': marginLeft === 'large+3' },
    { 'margin-left-large+4': marginLeft === 'large+4' },
    attributes.className,
  ]);

  return <div {...attributes} className={SpacerClassNames} />;
};

Spacer.propTypes = propTypes;
Spacer.defaultProps = defaultProps;

export default withBps({ mobileFirst: true })(Spacer);
