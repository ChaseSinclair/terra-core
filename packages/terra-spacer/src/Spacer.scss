@import './variables';
@import './mixins';

:local {
  /* stylelint-disable selector-class-pattern  */
  // constant value that cannot be themed
  .margin-none {
    margin: 0;
  }

  // themable values
  .margin-left-small-minus-2 {
    margin-left: var(--terra-spacer-margin-left-small-minus-2, 0.214rem); // 3px
  }

  .margin-left-small-minus-1 {
    margin-left: var(--terra-spacer-margin-left-small-minus-1, 0.357rem); // 5px
  }

  .margin-left-small {
    margin-left: var(--terra-spacer-margin-left-small, 0.5rem); // 7px
  }

  .margin-left-medium {
    margin-left: var(--terra-spacer-margin-left-medium, 0.714rem); // 10px
  }

  .margin-left-large {
    margin-left: var(--terra-spacer-margin-left-large, 0.857rem); // 12px
  }

  .margin-left-large-plus-1 {
    margin-left: var(--terra-spacer-margin-left-large-plus-1, 1.071rem); // 15px
  }

  .margin-left-large-plus-2 {
    margin-left: var(--terra-spacer-margin-left-large-plus-2, 1.429rem); // 20px
  }

  .margin-left-large-plus-3 {
    margin-left: var(--terra-spacer-margin-left-large-plus-3, 2.143rem); // 30px
  }

  .margin-left-large-plus-4 {
    margin-left: var(--terra-spacer-margin-left-large-plus-4, 3.571rem); // 50px
  }
  /* stylelint-enable */
}
