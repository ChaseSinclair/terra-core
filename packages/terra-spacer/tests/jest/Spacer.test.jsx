import React from 'react';
import Spacer from '../../src/Spacer';

describe('Spacer', () => {
  const defaultRender = <Spacer />;

  // Snapshot Tests
  it('should render a default component', () => {
    const wrapper = shallow(defaultRender);
    expect(wrapper).toMatchSnapshot();
  });

  // Prop Tests
  it('should render a spacer component with small top margin', () => {
    const wrapper = shallow(<Spacer marginTop="small">Test</Spacer>);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a spacer component with medium bottom margin', () => {
    const wrapper = shallow(<Spacer marginBottom="medium">Test</Spacer>);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a spacer component with large left padding', () => {
    const wrapper = shallow(<Spacer paddingLeft="large">Test</Spacer>);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a spacer component with inline-block display', () => {
    const wrapper = shallow(<Spacer isInlineBlock>Test</Spacer>);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a spacer component with right padding set to the given value at the given viewport size', () => {
    const wrapper = shallow(<Spacer bps={{ 768: { paddingRight: 'large+3' } }}>Test</Spacer>);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a spacer component with all margin props set to the given values', () => {
    const wrapper =
    shallow(
      <Spacer marginTop="small" marginBottom="small-1" marginLeft="small-2" marginRight="medium">
        Test
      </Spacer>);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a spacer component with all padding props set to the given values', () => {
    const wrapper =
    shallow(
      <Spacer paddingTop="large+1" paddingBottom="large+2" paddingLeft="large+3" paddingRight="large+4">
        Test
      </Spacer>);
    expect(wrapper).toMatchSnapshot();
  });

  // Custom Prop Test
  it('it should pass in a custom prop', () => {
    const wrapper = shallow(<Spacer className="spacer">Test</Spacer>);
    expect(wrapper).toMatchSnapshot();
  });
});
