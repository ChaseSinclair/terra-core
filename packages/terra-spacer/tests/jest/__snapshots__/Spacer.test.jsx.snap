// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spacer it should pass in a custom prop 1`] = `
<Spacer
  className="spacer"
  isInlineBlock={false}
  marginBottom="none"
  marginLeft="none"
  marginRight="none"
  marginTop="none"
  paddingBottom="none"
  paddingLeft="none"
  paddingRight="none"
  paddingTop="none"
>
  Test
</Spacer>
`;

exports[`Spacer should render a default component 1`] = `
<Spacer
  isInlineBlock={false}
  marginBottom="none"
  marginLeft="none"
  marginRight="none"
  marginTop="none"
  paddingBottom="none"
  paddingLeft="none"
  paddingRight="none"
  paddingTop="none"
/>
`;

exports[`Spacer should render a spacer component with all margin props set to the given values 1`] = `
<Spacer
  isInlineBlock={false}
  marginBottom="small-1"
  marginLeft="small-2"
  marginRight="medium"
  marginTop="small"
  paddingBottom="none"
  paddingLeft="none"
  paddingRight="none"
  paddingTop="none"
>
  Test
</Spacer>
`;

exports[`Spacer should render a spacer component with all padding props set to the given values 1`] = `
<Spacer
  isInlineBlock={false}
  marginBottom="none"
  marginLeft="none"
  marginRight="none"
  marginTop="none"
  paddingBottom="large+2"
  paddingLeft="large+3"
  paddingRight="large+4"
  paddingTop="large+1"
>
  Test
</Spacer>
`;

exports[`Spacer should render a spacer component with inline-block display 1`] = `
<Spacer
  isInlineBlock={true}
  marginBottom="none"
  marginLeft="none"
  marginRight="none"
  marginTop="none"
  paddingBottom="none"
  paddingLeft="none"
  paddingRight="none"
  paddingTop="none"
>
  Test
</Spacer>
`;

exports[`Spacer should render a spacer component with large left padding 1`] = `
<Spacer
  isInlineBlock={false}
  marginBottom="none"
  marginLeft="none"
  marginRight="none"
  marginTop="none"
  paddingBottom="none"
  paddingLeft="large"
  paddingRight="none"
  paddingTop="none"
>
  Test
</Spacer>
`;

exports[`Spacer should render a spacer component with medium bottom margin 1`] = `
<Spacer
  isInlineBlock={false}
  marginBottom="medium"
  marginLeft="none"
  marginRight="none"
  marginTop="none"
  paddingBottom="none"
  paddingLeft="none"
  paddingRight="none"
  paddingTop="none"
>
  Test
</Spacer>
`;

exports[`Spacer should render a spacer component with right padding set to the given value at the given viewport size 1`] = `
<Spacer
  bps={
    Object {
      "768": Object {
        "paddingRight": "large+3",
      },
    }
  }
  isInlineBlock={false}
  marginBottom="none"
  marginLeft="none"
  marginRight="none"
  marginTop="none"
  paddingBottom="none"
  paddingLeft="none"
  paddingRight="large+3"
  paddingTop="none"
>
  Test
</Spacer>
`;

exports[`Spacer should render a spacer component with small top margin 1`] = `
<Spacer
  isInlineBlock={false}
  marginBottom="none"
  marginLeft="none"
  marginRight="none"
  marginTop="small"
  paddingBottom="none"
  paddingLeft="none"
  paddingRight="none"
  paddingTop="none"
>
  Test
</Spacer>
`;
